utils.go

```diff
@@ -0,0 +1,193 @@
+// not externalized through the API. Hopefully in the future SelectHopHints
+// will be refactored to allow for custom data sources. It iterates through all
+// the active and public channels available and returns eligible channels.
+// Eligibility requirements are simple: does the channel have enough liquidity
+// to fulfill the request and is the node whitelisted (if specified)
+func SelectHopHints(ctx context.Context, lnd *lndclient.LndServices,
+	amtMSat btcutil.Amount, numMaxHophints int,
+	includeNodes map[route.Vertex]struct{}) ([][]zpay32.HopHint, error) {
+
+	// Fetch all active and public channels.
+	openChannels, err := lnd.Client.ListChannels(ctx, false, false)
+	if err != nil {
+		return nil, err
+	}
+
+	// We'll add our hop hints in two passes, first we'll add all channels
+	// that are eligible to be hop hints, and also have a local balance
+	// above the payment amount.
+	var totalHintBandwidth btcutil.Amount
+
+	// chanInfoCache is a simple cache for any information we retrieve
+	// through GetChanInfo
+	chanInfoCache := make(map[int]*lndclient.ChannelEdge)
+
+	// selectedCache is a simple cache which holds the indice of any
+	// channel we've added to final hopHints
+	skipCache := make(map[int]struct{})
+
+	hopHints := make([][]zpay32.HopHint, 0, numMaxHophints)
+
+	for i, channel := range openChannels {
+		// In this first pass, we'll ignore all channels in
+		// isolation can't satisfy this payment.
+
+		// Retrieve extra info for each channel not available in
+		// listChannels
+		chanInfo, err := lnd.Client.GetChanInfo(ctx, channel.ChannelID)
+		if err != nil {
+			return nil, err
+		}
+
+		// Cache the GetChanInfo result since it might be useful
+		chanInfoCache[i] = chanInfo
+
+		// Skip if channel can't forward payment
+		if channel.RemoteBalance < amtMSat {
+			log.Debugf(
+				"Skipping ChannelID: %v for hints as "+
+					"remote balance (%v sats) "+
+					"insufficientappears to be private",
+				channel.ChannelID, channel.RemoteBalance,
+			)
+			continue
+		}
+		// If includeNodes is set, we'll only add channels with peers in
+		// include_node. This is done to respect the last_hop parameter
+		if len(includeNodes) > 0 {
+			if _, ok := includeNodes[channel.PubKeyBytes]; !ok {
+				continue
+			}
+		}
+
+		// Mark the index to skip so we can skip it on the next
+		// iteration if needed. We'll skip all channels that make
+		// it past this point as they'll likely belong to private
+		// nodes or be selected
+		skipCache[i] = struct{}{}
+
+		// We want to prevent leaking private nodes, which we define as
+		// nodes with only private channels
+		//
+		// GetNodeInfo will never return private channels, even if
+		// they're somehow known to us. If there are any channels
+		// returned, we can consider the node to be public.
+		nodeInfo, err := lnd.Client.GetNodeInfo(
+			ctx, channel.PubKeyBytes, true,
+		)
+		if err != nil {
+			log.Warnf("Skipping ChannelID: %v for hints as peer "+
```

Comment:
Rather than all errors, could check grpc error code:

```
status, ok := status.FromError(err)
if ok && status.Code == code.NotFound{
     log warning
}else if err!=nil{
   return err
}
```

This way we only log a warning if we specifically get a not found error, and will fail out for other errors.